src/query/functions/linear/histogram_quantile_test.go

```diff
@@ -0,0 +1,283 @@
+			buckets: []indexedBucket{
+				{upperBound: 10, idx: 5},
+				{upperBound: math.Inf(-1), idx: 6},
+				{upperBound: 1, idx: 0},
+				{upperBound: 2, idx: 3},
+			},
+		},
+		`{no="infinity"}`: indexedBuckets{
+			buckets: []indexedBucket{
+				{upperBound: 0.1, idx: 4},
+				{upperBound: 0.2, idx: 14},
+				{upperBound: 0.3, idx: 114},
+			},
+		},
+		`{just="infinity"}`: indexedBuckets{
+			buckets: []indexedBucket{
+				{upperBound: math.Inf(1), idx: 4},
+			},
+		},
+		`{just="neg-infinity"}`: indexedBuckets{
+			buckets: []indexedBucket{
+				{upperBound: math.Inf(-1), idx: 4},
+			},
+		},
+	}
+
+	actual := bucketedSeries{
+		`{bar="baz"}`: indexedBuckets{
+			buckets: []indexedBucket{
+				{upperBound: 1, idx: 0},
+				{upperBound: 2, idx: 3},
+				{upperBound: 10, idx: 5},
+				{upperBound: math.Inf(1), idx: 6},
+			},
+		},
+	}
+
+	sanitizeBuckets(bucketed)
+	assert.Equal(t, actual, bucketed)
+}
+
+func TestBucketQuantile(t *testing.T) {
+	// single bucket returns nan
+	actual := bucketQuantile(0.5, []bucketValue{{upperBound: 1, value: 1}})
+	assert.True(t, math.IsNaN(actual))
+
+	// bucket with no infinity returns nan
+	actual = bucketQuantile(0.5, []bucketValue{
+		{upperBound: 1, value: 1},
+		{upperBound: 2, value: 2},
+	})
+	assert.True(t, math.IsNaN(actual))
+
+	actual = bucketQuantile(0.5, []bucketValue{
+		{upperBound: 1, value: 1},
+		{upperBound: math.Inf(1), value: 22},
+	})
+	assert.Equal(t, float64(1), actual)
+
+	// NB: tested against Prom
+	buckets := []bucketValue{
+		{upperBound: 1, value: 1},
+		{upperBound: 2, value: 2},
+		{upperBound: 5, value: 5},
+		{upperBound: 10, value: 10},
+		{upperBound: 20, value: 15},
+		{upperBound: math.Inf(1), value: 16},
+	}
+
+	actual = bucketQuantile(0, buckets)
+	assert.InDelta(t, float64(0), actual, 0.0001)
+
+	actual = bucketQuantile(0.15, buckets)
+	assert.InDelta(t, 2.4, actual, 0.0001)
+
+	actual = bucketQuantile(0.2, buckets)
+	assert.InDelta(t, float64(3.2), actual, 0.0001)
+
+	actual = bucketQuantile(0.5, buckets)
+	assert.InDelta(t, float64(8), actual, 0.0001)
+
+	actual = bucketQuantile(0.8, buckets)
+	assert.InDelta(t, float64(15.6), actual, 0.0001)
+
+	actual = bucketQuantile(1, buckets)
+	assert.InDelta(t, float64(20), actual, 0.0001)
+}
+
+func TestNewOp(t *testing.T) {
+	args := make([]interface{}, 0, 1)
+	_, err := NewHistogramQuantileOp(args, HistogramQuantileType)
+	assert.Error(t, err)
+
+	args = append(args, "invalid")
+	_, err = NewHistogramQuantileOp(args, HistogramQuantileType)
+	assert.Error(t, err)
+
+	args[0] = 2.0
+	_, err = NewHistogramQuantileOp(args, ClampMaxType)
+	assert.Error(t, err)
+
+	op, err := NewHistogramQuantileOp(args, HistogramQuantileType)
+	assert.NoError(t, err)
+
+	assert.Equal(t, HistogramQuantileType, op.OpType())
+	assert.Equal(t, "type: histogram_quantile", op.String())
+}
+
+func testQuantileFunctionWithQ(t *testing.T, q float64) [][]float64 {
+	args := make([]interface{}, 0, 1)
+	args = append(args, q)
```

Comment:
Just pass in the args instead of having to construct them here?

Reply:
Unfortunately it's pretty annoying to build a `[]interface{..}`, and that's what we need to pass into the function; since we only ever need the `q` value here, it should be fine to build the interface slice here